tokio-executor/src/thread_pool/builder.rs

```diff
@@ -0,0 +1,248 @@
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio_executor::thread_pool::Builder;
+    ///
+    /// let thread_pool = Builder::new()
+    ///     .stack_size(32 * 1024)
+    ///     .build();
+    /// ```
+    pub fn stack_size(&mut self, val: usize) -> &mut Self {
+        self.stack_size = Some(val);
+        self
+    }
+
+    /// Execute function `f` on each worker thread.
+    ///
+    /// This function is provided a function that executes the worker and is
+    /// expected to call it, otherwise the worker thread will shutdown without
+    /// doing any work.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio_executor::thread_pool::Builder;
+    ///
+    /// let thread_pool = Builder::new()
+    ///     .around_worker(|index, work| {
+    ///         println!("worker {} is starting up", index);
+    ///         work();
+    ///         println!("worker {} is shutting down", index);
+    ///     })
+    ///     .build();
+    /// ```
+    pub fn around_worker<F>(&mut self, f: F) -> &mut Self
+    where
+        F: Fn(usize, &mut dyn FnMut()) + Send + Sync + 'static,
+    {
+        self.around_worker = Some(Arc::new(Box::new(f)));
+        self
+    }
+
+    /// Create the configured `ThreadPool`.
+    ///
+    /// The returned `ThreadPool` instance is ready to spawn tasks.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio_executor::thread_pool::Builder;
+    ///
+    /// let thread_pool = Builder::new()
+    ///     .build();
+    /// ```
+    pub fn build(&self) -> ThreadPool {
+        self.build_with_park(|_| DefaultPark::new())
+    }
+
+    /// Create the configured `ThreadPool` with a custom `park` instances.
+    ///
+    /// The provided closure `build_park` is called once per worker and returns
+    /// a `Park` instance that is used by the worker to put itself to sleep.
+    pub fn build_with_park<F, P>(&self, mut build_park: F) -> ThreadPool
+    where
+        F: FnMut(usize) -> P,
+        P: Park + Send + 'static,
+    {
+        let (shutdown_tx, shutdown_rx) = shutdown::channel();
+
+        let (pool, workers) = worker::create_set(self.pool_size, |i| BoxedPark::new(build_park(i)));
+
+        // Spawn threads for each worker
+        for (idx, mut worker) in workers.into_iter().enumerate() {
+            let around_worker = self.around_worker.clone();
+            let shutdown_tx = shutdown_tx.clone();
+
+            let mut th = thread::Builder::new().name(format!("{}{}", self.name_prefix, idx));
+
+            if let Some(stack) = self.stack_size {
+                th = th.stack_size(stack);
+            }
+
+            let res = th.spawn(move || {
+                struct AbortOnPanic;
+
+                impl Drop for AbortOnPanic {
+                    fn drop(&mut self) {
+                        if thread::panicking() {
+                            eprintln!("[ERROR] unhandled panic in Tokio scheduler");
```

Comment:
Should this message include that this is likely a bug in Tokio and should be reported?